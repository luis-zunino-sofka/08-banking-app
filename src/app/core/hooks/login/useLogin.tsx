import { ROUTE_PATH } from "@core/constants";
import { ILoginRequest, ILoginResponse } from "@core/interfaces";
import { login } from "@core/services/login";
import { encryptAES } from "@core/utils";
import { useForm } from "react-hook-form";
import toast from "react-hot-toast";
import { useNavigate } from "react-router-dom";
import { localStorageProperties } from "../../constants/localStoragesProperties";
import { useState } from "react";

interface LoginFormData {
  username: string;
  password: string;
}
export const useLogin = (
  saveLoginData: (loginData: ILoginResponse) => void
) => {
  const [isLoading, setIsLoading] = useState(false);
  const navigate = useNavigate();
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<LoginFormData>();

  const callLogin = async (data: ILoginRequest) => {
    setIsLoading(true);
    try {
      setIsLoading(true);

      const result = await login(data);
      if (result.dinBody) {
        // TODO: mejorar esto
        localStorage.setItem(
          localStorageProperties.customerId,
          String(result?.dinBody?.customerId)
        );
        localStorage.setItem(
          localStorageProperties.token,
          String(result?.dinBody?.token)
        );

        saveLoginData(result.dinBody);
        navigate(ROUTE_PATH.DASHBOARD);
      } else {
        toast(result.dinError.detail);
      }
      return result;
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      toast("Ha ocurrido un error ");
    } finally {
      setIsLoading(false);
    }
  };

  const onSubmit = (data: LoginFormData) => {
    callLogin({
      username: data.username,
      encryptedPassword: encryptAES(data.password),
    });
    localStorage.setItem(
      localStorageProperties.username,
      String(data.username)
    );
  };

  return { isLoading, errors, onSubmit, register, handleSubmit };
};
