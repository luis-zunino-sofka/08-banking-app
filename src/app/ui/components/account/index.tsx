import { useCreateBankAccount } from "@core/hooks";
import { decryptAES } from "@core/utils";
import React from "react";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";
import "./styles.scss";

export const CreateBankAccount: React.FC = () => {
  const { isLoading, accounts, errors, handleSubmit, register, onSubmit } =
    useCreateBankAccount();

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmit)} className="form">
        <div style={{ display: "flex", gap: "10px" }}>
          <div>
            <div>
              <label htmlFor="accountType">Tipo de Cuenta</label>
              <select
                id="accountType"
                {...register("accountType", {
                  required: "Selecciona un tipo de cuenta",
                })}
              >
                <option key={1} value="">
                  Selecciona el tipo de cuenta
                </option>
                <option key={2} value="corriente">
                  Corriente
                </option>
              </select>
              <span className="error-message">
                {errors.accountType && errors.accountType.message}
              </span>
            </div>
            <div>
              <label htmlFor="initialBalance">Saldo Inicial</label>
              <input
                type="number"
                id="initialBalance"
                placeholder="Saldo inicial en $"
                min={0}
                {...register("initialBalance", {
                  required: "El saldo inicial es obligatorio",
                  min: {
                    value: 0,
                    message: "El saldo inicial no puede ser negativo",
                  },
                })}
              />
              <span className="error-message">
                {errors.initialBalance && errors.initialBalance.message}
              </span>
            </div>
            <Button label="Crear Cuenta" isLoading={isLoading} />
          </div>
          <div className="user-accounts">
            <p>Cuentas del usuario:</p>
            {accounts.map((a) => (
              <div key={a.accountId} className="user-accounts__item">
                <div className="user-accounts__info">
                  <span>NÂº de cuenta:</span> {decryptAES(a.encryptedNumber)}
                </div>
                <div className="user-accounts__info amount">
                  <span>Monto:</span> $ {a.amount}
                </div>
              </div>
            ))}
          </div>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
