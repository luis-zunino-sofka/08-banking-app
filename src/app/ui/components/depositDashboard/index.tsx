import { useDeposit } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";

export const DepositDashboard = () => {
  const {
    isLoading,
    accounts,
    errors,
    isAccountSelected,
    handleSubmit,
    register,
    onSubmit,
  } = useDeposit();

  return (
    <div className="deposit">
      <form
        id="deposit-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <label htmlFor="deposit-source">
          Selecciona el origen del depósito:
        </label>
        <select
          id="deposit-source"
          {...register("depositSource", {
            required: "El monto a depositar es obligatorio",
          })}
        >
          <option value="branch-deposit" defaultValue="branch-deposit">
            Depósito desde sucursal (Sin costo)
          </option>
          <option value="atm">
            Depósito desde cajero automático ($2 de costo)
          </option>
          <option value="account">
            Depósito desde otra cuenta ($1.5 de costo)
          </option>
        </select>
        {errors.depositSource && (
          <span className="error-message">{errors.depositSource.message}</span>
        )}
        <label htmlFor="deposit-amount">Monto:</label>
        <input
          type="number"
          id="deposit-amount"
          placeholder="Monto a depositar"
          required
          min={0}
          {...register("amount", {
            required: "El monto a depositar es obligatorio",
            min: {
              value: 1,
              message: "El monto a depositar debe ser mayor a 0",
            },
          })}
        />
        {errors.amount && (
          <span className="error-message">{errors.amount.message}</span>
        )}
        <div>
          <label htmlFor="accountId">
            {isAccountSelected
              ? "Cuenta de extracción:"
              : "Cuenta a depositar:"}
          </label>
          <select
            id="accountId"
            {...register("accountId", {
              required: "La cuenta bancaria es obligatorio",
            })}
            required
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountNumberToDeposit && (
            <span className="error-message">
              {errors.accountNumberToDeposit.message}
            </span>
          )}
        </div>
        {isAccountSelected && (
          <>
            <label htmlFor="account-number-to-deposit">
              Selecciona la cuenta a depositar:
            </label>
            <input
              type="text"
              id="account-number-to-deposit"
              placeholder="Cuenta a depositar"
              required
              min={0}
              {...register("accountNumberToDeposit", {
                required: "La cuenta a depositar es obligatorio",
              })}
            />

            {errors.accountNumberToDeposit && (
              <span className="error-message">
                {errors.accountNumberToDeposit.message}
              </span>
            )}
          </>
        )}
        <Button label="Depositar" isLoading={isLoading} />
        <p>* El costo de la transacción se deducirá del saldo.</p>
      </form>
      <Toaster />
    </div>
  );
};
