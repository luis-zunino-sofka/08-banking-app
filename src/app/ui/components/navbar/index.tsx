import "./styles.scss";
import { INavItem, IState } from "@core/interfaces";
import { FC } from "react";

export interface IDashboardProps {
  state: IState;
  isRefechingAccounts: boolean;
  navItems: INavItem[];
  refetchAccounts: () => void;
  handleLogout: () => void;
}

export const Navbar: FC<IDashboardProps> = (props) => {
  const {
    state,
    navItems,
    isRefechingAccounts,
    refetchAccounts,
    handleLogout,
  } = props;

  return (
    <nav className="navbar">
      <div className="navbar__title">
        <h1>Bienvenido {state.loginData?.username}</h1>
      </div>
      <div className="navbar__balance">
        <span>Saldo disponible: </span>
        <strong>${state.balance}</strong>
        <button
          disabled={isRefechingAccounts}
          className="navbar__reload"
          onClick={() => refetchAccounts()}
        >
          Actualizar
        </button>
      </div>
      <div className="navbar__nav">
        <ul className="navbar__list">
          {navItems.map((item) => (
            <li key={item.path} className="navbar-item">
              <button
                className={`navbar__link ${
                  item.isActive ? "navbar__link--active" : ""
                }`}
                onClick={item.onClick}
              >
                {item.label}
              </button>
            </li>
          ))}
          <li className="navbar__item">
            <button
              className="navbar__link navbar__logout"
              onClick={handleLogout}
            >
              Cerrar sesi√≥n
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
