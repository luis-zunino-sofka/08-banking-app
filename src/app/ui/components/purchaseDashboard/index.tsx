import { Toaster } from "react-hot-toast";
import { usePurchase } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Button } from "../button";
import "../../styles/styles.scss";

export const PurchaseDashboard = () => {
  const { isLoading, accounts, errors, register, handleSubmit, onSubmit } =
    usePurchase();

  return (
    <div>
      <form
        id="purchase-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div>
          <label htmlFor="purchase-type">Selecciona el tipo de compra:</label>
          <select
            id="purchase-type"
            {...register("purchaseType", {
              required: "Por favor selecciona un tipo de compra",
            })}
          >
            <option value="">Elige el tipo de compra</option>
            <option value="physical">Establecimiento físico (Sin costo)</option>
            <option value="online">En línea (Costo de seguro de $5 USD)</option>
          </select>
          {errors.purchaseType && (
            <span className="error-message">{errors.purchaseType.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="account-number-to-deposit">
            Selecciona la cuenta a depositar:
          </label>
          <select
            id="account-number-to-deposit"
            {...register("accountId", {
              required: "La cuenta bancaria a depositar es obligatorio",
            })}
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountId && (
            <span className="error-message">{errors.accountId.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="purchase-amount">
            Ingresa el monto a depositar:
            <input
              type="number"
              id="purchase-amount"
              placeholder="Monto de la compra en $"
              {...register("purchaseAmount", {
                required: "El monto es obligatorio",
                min: {
                  value: 1,
                  message: "El monto debe ser mayor que $0",
                },
              })}
            />
          </label>
          {errors.purchaseAmount && (
            <span className="error-message">
              {errors.purchaseAmount.message}
            </span>
          )}
        </div>
        <Button label="Comprar" isLoading={isLoading} />
      </form>

      <Toaster />
    </div>
  );
};
