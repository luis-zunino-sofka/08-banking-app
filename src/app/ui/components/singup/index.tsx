import { useSignup } from "@core/hooks/signup/useSignup";
import { ROUTE_PATH } from "@core/constants";
import { Toaster } from "react-hot-toast";
import "./styles.scss";
import { InputSignup } from "./InputSignup";

export const SignupForm = () => {
  const { errors, isLoading, onSubmit, register, handleSubmit, watch } =
    useSignup();

  return (
    <div className="signup-form">
      <form
        id="register-form"
        className="signup-form__form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2 className="signup-form__title">Regístrate</h2>
        <div className="flex">
          <InputSignup
            register={register("name", {
              required: "El nombre es obligatorio",
            })}
            name="name"
            type="text"
            label="Nombre:"
            errors={errors}
            placeHolder="Ingrese su nombre"
          />

          <InputSignup
            register={register("lastName", {
              required: "El apellido es obligatorio",
            })}
            name="lastName"
            type="text"
            label="Apellido:"
            errors={errors}
            placeHolder="Ingrese su apellido"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("dni", { required: "El DNI es obligatorio" })}
            name="dni"
            type="text"
            label="DNI:"
            errors={errors}
            placeHolder="Ingrese su dni"
          />

          <InputSignup
            register={register("username", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="username"
            type="text"
            label="Nombre de usuario:"
            errors={errors}
            placeHolder="Ingrese su nombre de usuario"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("password", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="password"
            type="password"
            label="Contraseña:"
            errors={errors}
            placeHolder="Contraseña"
          />

          <InputSignup
            register={register("confirmPassword", {
              required: "La confirmación de la contraseña es obligatoria",
              validate: (value) =>
                value === watch("password") || "Las contraseñas no coinciden",
            })}
            name="confirmPassword"
            type="password"
            label="Confirmar contraseña:"
            errors={errors}
            placeHolder="Confirmar contraseña"
          />
        </div>
        <button
          type="submit"
          className="signup-form__submit"
          disabled={isLoading}
        >
          Registrarse
        </button>
        <div className="signup-form__footer">
          <p>
            ¿Ya tienes cuenta?
            <a
              href={ROUTE_PATH.LOGIN}
              id="login-link"
              className="signup-form__login-link"
            >
              Inicia sesión aquí
            </a>
          </p>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
