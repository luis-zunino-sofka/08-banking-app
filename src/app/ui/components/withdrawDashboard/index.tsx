import { Toaster } from "react-hot-toast";
import { decryptAES } from "@core/utils";
import { useWithdrawDashboard } from "@core/hooks";

export const WithdrawDashboard = () => {
  const { accounts, register, handleSubmit, errors, onSubmit } =
    useWithdrawDashboard();

  return (
    <div className="withdraw">
      <form
        id="withdraw-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div>
          <label htmlFor="withdraw-amount">Monto:</label>
          <input
            type="number"
            id="withdraw-amount"
            placeholder="Monto a retirar en $"
            {...register("amount", {
              required: "El monto es obligatorio.",
              min: {
                value: 0,
                message: "El monto debe ser mayor que 0.",
              },
            })}
          />
          {errors.amount && (
            <span className="error-message">{errors.amount.message}</span>
          )}
          <>
            <label htmlFor="account-number-to-withdraw">
              Selecciona la cuenta a retirar:
            </label>
            <select
              id="account-number-to-withdraw"
              {...register("accountNumberToWithdraw", {
                required: "La cuenta bancaria a retirar es obligatorio",
              })}
              required
            >
              <option value="">Elige una cuenta</option>
              {accounts.map((a) => (
                <option key={a.accountId} value={a.accountId}>
                  Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
                </option>
              ))}
            </select>

            {errors.accountNumberToWithdraw && (
              <span className="error-message">
                {errors.accountNumberToWithdraw.message}
              </span>
            )}
          </>
        </div>

        <p>* El retiro en cajero tiene un costo de $1 USD por transacci√≥n.</p>
        <button type="submit">Retirar</button>
      </form>
      <Toaster />
    </div>
  );
};
