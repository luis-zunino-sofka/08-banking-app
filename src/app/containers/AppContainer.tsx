import { useNavbar } from "@core/hooks";
import { useAccountContext, useAppContext } from "@core/state";
import { Navbar } from "@ui/components/navbar";
import { AppLayout } from "@ui/layouts";
import React from "react";
import { Toaster } from "react-hot-toast";

export const AppContainer = ({ children }: { children: React.ReactNode }) => {
  const { state } = useAppContext();
  const { refetchAccounts } = useAccountContext();
  const { navItems, handleLogout } = useNavbar();

  return (
    <AppLayout>
      <Navbar
        state={state}
        handleLogout={handleLogout}
        refetchAccounts={refetchAccounts}
        navItems={navItems}
      />
      {children}
      <Toaster />
    </AppLayout>
  );
};
