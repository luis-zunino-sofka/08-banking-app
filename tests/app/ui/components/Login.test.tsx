import {
  afterEach,
  beforeEach,
  describe,
  expect,
  Mock,
  test,
  vi,
} from "vitest";
import { fireEvent, render, waitFor, screen } from "@testing-library/react";
import React from "react";
import "@testing-library/jest-dom";
import { Login } from "../../../../src/app/ui/components/login";
import { AccountContextMock } from "../containers/AppContainer.test";
import { mockAccountState, mockAppState } from "../../../mocks/dataMocked";
import { encryptAES } from "../../../../src/app/core/utils/encryptAES.utils";

vi.mock("@core/hooks", () => ({
  useLogin: vi.fn(),
}));

const Wrapper = () => (
  <AccountContextMock
    mockAppState={mockAppState}
    children={<Login saveLoginData={mockSaveLoginData} />}
    mockAccountState={mockAccountState}
  />
);

let mockSaveLoginData: Mock;
let mockOnSubmit: Mock;
let mockHandleSubmit: Mock;

describe("Login Component", () => {
  beforeEach(() => {
    mockSaveLoginData = vi.fn();
    mockOnSubmit = vi.fn();
    mockHandleSubmit = vi.fn();

    vi.mock("@core/hooks", () => ({
      useSignup: () => ({
        isLoading: false,
        errors: {
          username: { message: "El nombre es obligatorio" },
          password: { message: "La contraseña es obligatoria" },
        },
        onSubmit: mockOnSubmit,
        register: vi.fn(),
        handleSubmit: mockHandleSubmit,
      }),
    }));

    render(<Wrapper />);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  test("Match snapshoot", () => {
    const { asFragment } = render(<Wrapper />);
    expect(asFragment()).toMatchSnapshot();
  });

  test(" debe renderizar el formulario correctamente", () => {
    screen.getByRole("heading", { level: 2, name: /^Iniciar sesión$/i });
    screen.getByLabelText(/Nombre de usuario/i);
    screen.getByLabelText(/Contraseña/i);
    screen.getByRole("button", { name: /^Iniciar sesión$/i });
    screen.getByRole("link", { name: /Regístrate aquí/i });
    screen.getByText(/¿No tienes cuenta?/i);
  });

  test(" debe mostrar errores de validación", async () => {
    const submitButton = screen.getByRole("button", {
      name: /^Iniciar sesión$/i,
    });
    const nombre = screen.getByPlaceholderText(/^Nombre de usuario$/i);
    const password = screen.getByPlaceholderText(/^Contraseña$/i);

    fireEvent.change(nombre, {
      target: { value: "Pepe" },
    });

    fireEvent.change(password, {
      target: { value: "1" },
    });

    await waitFor(() => fireEvent.click(submitButton));

    expect(nombre).toHaveValue("Pepe");
    expect(password).toHaveValue("1");

    await waitFor(() =>
      screen.getByText("La contraseña debe tener al menos 6 caracteres")
    );

    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  test("debe llamar a onSubmit cuando el formulario es válido", async () => {
    const usernameInput = screen.getByLabelText("Nombre de usuario:");
    const passwordInput = screen.getByLabelText("Contraseña:");
    const submitButton = screen.getByRole("button", {
      name: /^Iniciar sesión$/i,
    });

    waitFor(() => {
      fireEvent.change(usernameInput, { target: { value: "juanperez" } });
      fireEvent.change(passwordInput, {
        target: { value: encryptAES("123456") },
      });
    });

    expect(usernameInput).toHaveValue("juanperez");
    expect(passwordInput).toHaveValue("diaFFu/2IHVSQj17PI1amg==");

    waitFor(() => fireEvent.click(submitButton));
    waitFor(() => expect(mockOnSubmit).toHaveBeenCalled());
  });

  test(" debe llamar a saveLoginData al hacer submit exitoso", async () => {
    const usernameInput = screen.getByLabelText("Nombre de usuario:");
    const passwordInput = screen.getByLabelText("Contraseña:");
    const submitButton = screen.getByRole("button", {
      name: /^Iniciar sesión$/i,
    });

    const mockfn = vi.fn().mockImplementation(encryptAES);
    waitFor(() => {
      fireEvent.change(usernameInput, { target: { value: "juanperez" } });
      fireEvent.change(passwordInput, {
        target: { value: mockfn("123456") },
      });
    });

    await waitFor(() => fireEvent.click(submitButton));

    expect(usernameInput).toHaveValue("juanperez");
    expect(passwordInput).toHaveValue("diaFFu/2IHVSQj17PI1amg==");
    waitFor(() => expect(mockOnSubmit).toBeCalled());
  });

  test('debe redirigir al registro cuando se hace clic en "Regístrate aquí"', () => {
    const registerLink = screen.getByRole("link", { name: /Regístrate aquí/i });
    expect(registerLink).toHaveAttribute("href", "/sign-up");
  });
});
