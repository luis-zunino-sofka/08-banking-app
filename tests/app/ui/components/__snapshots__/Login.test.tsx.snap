// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountProvider > Match snapshoot 1`] = `<DocumentFragment />`;

exports[`Login Component > Match snapshoot 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <form
      class="form"
      id="login-form"
    >
      <h2
        class="login__form-title"
      >
        Iniciar sesión
      </h2>
      <div
        class="login__input-group"
      >
        <label
          for="username"
        >
          Nombre de usuario:
        </label>
        <input
          id="username"
          name="username"
          placeholder="Nombre de usuario"
          required=""
          type="text"
        />
      </div>
      <div
        class="login__input-group"
      >
        <label
          for="password"
        >
          Contraseña:
        </label>
        <input
          id="password"
          name="password"
          placeholder="Contraseña"
          required=""
          type="password"
        />
      </div>
      <button
        type="submit"
      >
        Iniciar sesión
      </button>
      <p
        class="login__register-link"
      >
        ¿No tienes cuenta?
        <a
          href="/sign-up"
          id="register-link"
        >
          Regístrate aquí
        </a>
      </p>
    </form>
  </div>
</DocumentFragment>
`;
